### DISCUSSION TITLE
help with log_field

### DISCUSSION THREAD

**Discussion by rsdavies** [MetOffice/simulation-systems#117] [LFRic]
Hi All, 
I'm debugging some issues, and in lieu of being able to get a proper debugging session going, I am logging out various fields which I think are causing my issue, e.g.

`CALL tile_snow_mass%log_field(LOG_LEVEL_INFO, "tile_snow_mass")`

I have a few questions: 
 1. Can I get it to log out different layers/levels/pseudolevels?
 2. Can I assume that what comes out in PET0.log... this time, will be the same as what comes out next time?  
 3. I've been previously told 'anything beyond line 64 is junk'. Is this always the case, or does it rely on something else?
 
Thanks in advance!

**Comment by stevemullerworth** [comment]
Hi Rhiannon,
1. No. But the log_field code is not rocket science if you are feeling brave. It should be easy to add a couple of optional arguments and make the required changes to enable you to specify a particular layer and/or df. The df argument would relate to the elements of the pseudolevels.

https://code.metoffice.gov.uk/trac/lfric/browser/LFRic/trunk/infrastructure/source/field/field_r64_mod.f90?rev=47735#L630

2. The model should be bit reproducible when rerun with the same number of ranks and same number of threads.
3. Not sure I understand the Line 64 comment. If you are running a configuration where each rank owns 64 cells (e.g. a C8 on 6 ranks or a C16 on 24 ranks) then you will still see output from the halo cells beyond cell 64 which may contain junk. Is that an explanation?

**Reply by rsdavies** [reply]
Well the answer for 1 is what I feared, but I can probably give it a stab. 

2. I'm glad that its always going to log the same bit of the field into the same log each time, that certainly helps. 
3. That is exactly what I wanted to know, though how does one know how many cells a rank is owning?

**Comment by stevemullerworth** [comment]
For a Cn mesh running on m ranks it is 6n^2/m.
If it's a LAM mesh it's a bit more complicated. If the mesh divides evenly into the choice of partitioning, it is straightforward. If there are any remainder rows/columns after dividing the mesh, then they are added from left to right/top to bottom. If I had to guess, PE numbering goes from left to right and from top to bottom...

**Reply by rsdavies** [reply]
I have a suspicion something is off with my reading in this file, I've set all the vales to be 0 for the first layer, 1 for the second layer etc, and its logged me out
![image](https://github.com/MetOffice/simulation-systems/assets/26064197/4debf871-1c1a-4475-8122-bc9b23c05937)
Which I was not expecting! I figured I'd get all 0s...

**Reply by stevemullerworth** [reply]
Looks like a field with a single layer and no multidata.

Can you point to code on a branch (or a working copy in your home space)? Happy to dig in and communicate via email/MS Teams till we get to the bottom of the issue.

**Reply by rsdavies** [reply]
Hi Steve, 
See branch at https://code.metoffice.gov.uk/trac/lfric/log/LFRic/branches/dev/rhiannondavies/r44221_surf_ancil_files

I'd love to know what I've done wrong! No doubt something ridiculous.
